In [2]:
import sys

d = {}
prev_size = sys.getsizeof(d)

for i in range(1000):
    d[i] = i  # 添加元素
    new_size = sys.getsizeof(d)
    if new_size != prev_size:
        print(f"元素数: {i+1} → 内存变化: {prev_size} → {new_size} 字节")
        prev_size = new_size

元素数: 1 → 内存变化: 64 → 224 字节
元素数: 6 → 内存变化: 224 → 352 字节
元素数: 11 → 内存变化: 352 → 632 字节
元素数: 22 → 内存变化: 632 → 1168 字节
元素数: 43 → 内存变化: 1168 → 2264 字节
元素数: 86 → 内存变化: 2264 → 4688 字节
元素数: 171 → 内存变化: 4688 → 9304 字节
元素数: 342 → 内存变化: 9304 → 18512 字节
元素数: 683 → 内存变化: 18512 → 36952 字节


In [5]:
import sys

def check_dict_resize():
    d = {}
    sizes = []
    for i in range(100):
        d[i] = i
        if not sizes or sys.getsizeof(d) != sizes[-1][-1]:
            sizes.append((i+1, sys.getsizeof(d)))
    return sizes

print(check_dict_resize())

[(1, 224), (6, 352), (11, 632), (22, 1168), (43, 2264), (86, 4688)]


字典并不是每插入一个键就立即变大，而是当装载因子(load factor)超过 2/3 时，把桶数组(bucket array)一次性扩容到原来的 2 倍左右。这就是你看到的 (元素个数, 字典对象总大小) 跳跃式增长的表： 

[(1, 224), (6, 352), (11, 632), (22, 1168), (43, 2264), (86, 4688)]